app/models/caffeinate/mailing.rb

```diff
@@ -131,7 +131,7 @@ def deliver_later!
     end
 
     def end_if_no_mailings!
-      end! if future_mailings.empty?
+      caffeinate_campaign_subscription.end! if caffeinate_campaign_subscription.future_mailings.empty?
```

Comment:
I think we might want to think about this one a little more.

When a campaign subscription is _ended_ it means that the same subscriber could be subscribed again using the plain `FooDripper.subscribe @thing` API — "ended" = "it'll behave as if it was never subscribed in the first place". That's strictly different from "completed," where `end!` was never expressly called, the campaign subscription just got through all its mailings. That's encoded in the campaign subscription:

```ruby
# app/models/caffeinate/campaign_subscription.rb
    def completed?
      caffeinate_mailings.unsent.count.zero?
    end
```

And when a campaign subscription is _completed_ — a subsequent call to `FooDripper.subscribe @thing` will return _that completed subscription_; not scaffold up a new subscription.

And, FWIW, I like that paradigm and difference myself: to "end" means to give room for a new one (to fully abort), to "complete", and even to "unsubscribe", is to recognize the presence of a prior subscription and _not_ give room for a new one.

So coming back to this mailing callback, I think we may _not_ want to do `caffeinate_campaign_subscription.end! if caffeinate_campaign_subscription.future_mailings.empty?` because it leaves no room for `completed?` to accurately represent its cause. This line implies that a campaign subscription will never be completed _without_ being ended too. I don't think that's the contract we want to set forth 🤔

Reply:
I realize that `end_if_no_mailings!` is rarely called since it's specifically for when `.touch`ed and doesn't run on updates or otherwise, but still I think it's a good discussion to figure out the expectations around "end" vs. "complete" etc.

Reply:
@joshmn I came back to this while doing the rounds on our app updating dependencies and saw that this issue was released in 2.5.0. Did you ever have a chance to mull over the above comments? I'm going to keep our app at 2.4.0 since we very actively differentiate between 'ended' and 'just out of mailings' 🤔

Reply:
Circling back here 2 days later; I see / realize now that this is really just here for testing purposes (and is only used internally in one place; a spec). I do think our app can go ahead and update now — we're certainly not running `#touch` manually on mailing records.. and frankly I doubt others are either... so that means the general contract of "`#end!` == a new subscription will be made" vs. "running out of mailings _or_ unsubscribing == won't make a new subscription on the next subscribe call" is still valid and upheld

👍